packages/rnpm-plugin-windows/src/common.js

```diff
@@ -41,61 +41,88 @@ function getLatestVersion() {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(rel[0].version);
+        }
+      });
+    } else {
+      npm.packages.releases('react-native-windows', (err, rels) => {
+        if (err) {
+          reject(err);
+        } else {
+          const matchingVersions = Object.keys(rels)
+            .filter(v => isVersionMatch(v, versionRange, tag))
+            .sort(semver.rcompare);
 
-      const matchedVersion = release.version;
-      const matchedPrerelease = semver.prerelease(matchedVersion);
-      const isPrerelease = tag && !!matchedPrerelease;
-      if (!isVersionMatch(matchedVersion, version, tag) && (ignoreStable || isPrerelease)) {
-        const versions = Object.keys(release.versions);
-        const candidates = versions.filter(v => isVersionMatch(v, version, tag)).sort(semver.rcompare);
-        if (candidates.length === 0) {
-          const tagMatches = versions.filter(v => isTagMatch(v, tag)).sort(semver.rcompare);
-          if (tagMatches.length === 0) {
-            reject(new Error(`Could not find react-native-windows@${version}-${tag}.*.`));
+          if (matchingVersions.length > 0) {
+            resolve(matchingVersions[0]);
           } else {
-            const latestVersion = tagMatches[0];
-            reject(new Error(`Could not find react-native-windows@${version}-${tag}.*. ` +
-              `Latest version of react-native-windows for tag '${tag}' is ${latestVersion}, try switching to ` +
-              `react-native@${semver.major(latestVersion)}.${semver.minor(latestVersion)}.*.`));
+            reject();
           }
         }
-        resolve(candidates[0]);
-      } else {
-        resolve(matchedVersion);
-      }
-    });
+      });
+    }
   });
 }
 
-const getInstallPackage = function (version, tag, useStable) {
+async function getMatchingVersion(versionRange, tag) {
+  const versionStr = tag === 'master' ? 'master' : versionRange;
+  console.log(`Checking for react-native-windows version matching ${versionStr}...`);
+
+  try {
+    return await getLatestMatchingVersion(versionRange, tag);
+  } catch (ex) {
+    const latestVersion = await getLatestVersion();
+    throw new Error(`Could not find react-native-windows@${versionStr}. ` +
+      `Latest version of react-native-windows is ${latestVersion}, try switching to ` +
+      `react-native@${semver.major(latestVersion)}.${semver.minor(latestVersion)}.*.`);
+  }
+}
+
+const isSupportedVersion = function(validVersion, validRange) {
+  // Allow 0.0.0-x master builds
+  if (validVersion) {
+    return semver.lt(validVersion, '0.0.0') || semver.gte(validVersion, '0.27.0');
+  } else if (validRange) {
+    return semver.gtr('0.0.0', validRange) || semver.ltr('0.27.0', validRange);
+  } else {
+    return false;
+  }
+};
+
+const getInstallPackage = function (version, tag) {
   const packageToInstall = 'react-native-windows';
   const validVersion = semver.valid(version);
   const validRange = semver.validRange(version);
```

Comment:
>const validRange = semver.validRange(version); [](start = 2, length = 46)

I'm trying to see how this will ever return a range - IIUC, version must be a range for validRange to do anything.  But version never seems to be a range.  Am I missing something?

Reply:
`version` being a range is actually the common case here. `getReactNativeVersion` unintuitively returns something like `0.61.*` instead of a version. We also accept input from the `--windowsVersion` arg though, where people can specifiy a range.

Reply:
Thanks I see it now.  Sorry if I'm being obtuse but I don't follow the logic here.  The user installed a specific version of react-native, but we seem to be ignoring the patch portion of that version.  We then move on to see if we have a matching version of react-native-windows.  How does a range for patch versions of react-native help us find a compatible version of react-native-windows?

Maybe there is a reason but I've been staring at this code for about an hour now.  Some comments would surely help!

---
In reply to: [397415096](https://github.com/microsoft/react-native-windows/pull/4403#discussion_r397415096) [](ancestors = 397415096)

Reply:
I can definitely add some comments. See the last email thread where I mentioned existing version matching logic is a bit of a hot mess 😛 

There are a few scenarios:

One is that someone specifies an exact version of RNW using the `windowsVersion` arg. We don't run matching logic at all in this case and just use the argument as verbatim.

Two is that no argument is specified, and a non-master template is given. In this case logic is to grab the current RN version and try to match its major and minor versions for our RNW package as something like `0.61.*`. This gets a bit funky though, since that range won't actually pick up anything with tags, since those are prereleases. There is instead custom logic to basically just enumerate all packages and find the newest which matches major, minor, and tag instead of querying for the range directly. This same logic is used for the no tag case.

There was previously some fuzziness to the logic here where tags didn't actually have to match if none was given. That isn't really what we want, since we want preview to be explicitly opt-in. We were by chance avoiding that logic before since we were forcing `vnext` or `legacy` for recent versions. I ended up removing this fuzzy logic entirely and just hardcoding the right tags to match against for legacy versions to preserve behavior.

Three is if we have a `master` tag. I.e. if someone adds `--template master`. Here I ignore RN version/explicit version range arguments entirely and just get the latest tagged as master, since we don't have the association between RN versions.

Four is if someone passes in a range for `windowsVersion`. There is some special logic in checking for version matches along with tag for this, but really, since we are doing the tag shenanigans, we are not respecting the range passed in entirely. I have some doubts the existing logic is doing the right thing here in a lot of cases, but didn't spend a lot of time trying to fix this.

Reply:
Thanks for the detailed explanation (this is pretty close the comments I'd like to see in the code).  As much as it pains me to see so much complexity in the algorithm, I think this is probably necessary and good enough for now.

---
In reply to: [397438193](https://github.com/microsoft/react-native-windows/pull/4403#discussion_r397438193) [](ancestors = 397438193)